### PR TITLE
cmake: allow SwiftSyntax parser to have an externally defined version number in the dylib

### PR DESCRIPTION
**Pull request by nkcsgexi**

### CODE REVIEW

File: tools/libSwiftSyntaxParser/CMakeLists.txt [hunk 1]
```diff
@@ -20,6 +20,13 @@ add_llvm_symbol_exports(libSwiftSyntaxParser ${LLVM_EXPORTED_SYMBOL_FILE})
 # Adds -dead_strip option
 add_link_opts(libSwiftSyntaxParser)
 
+if(LIBPARSER_VER)
+  set(LIBPARSER_LINK_FLAGS " -Wl,-compatibility_version -Wl,1")
```
Comment (akyrtzi) [inline tools/libSwiftSyntaxParser/CMakeLists.txt]: 2 improvements:
* Add `SWIFT_` prefix to the CMake variable
* Those link flags are only valid for macOS, add handling of the version for other platforms by following what libclang is doing here: https://github.com/apple/swift-clang/blob/38ef19dd1f3ea3d09a8b7e77b29aa95fd6ad7d87/tools/libclang/CMakeLists.txt#L108

**Review by akyrtzi — CHANGES_REQUESTED** [review]

File: tools/libSwiftSyntaxParser/CMakeLists.txt (hunk 1 above)
Comment (nkcsgexi) [reply tools/libSwiftSyntaxParser/CMakeLists.txt]: Good point! Done. 
Do we need  `DEFINE_SYMBOL _CINDEX_LIB_` here? does SwiftSyntax parser have a similar symbol?

Comment (akyrtzi) [reply tools/libSwiftSyntaxParser/CMakeLists.txt]: No you can ignore this, for the parser we just leave it to CMake to automatically define the right macro, which is then checked here: https://github.com/apple/swift/blob/b5299531d4e498002920ebb0ddf853e12212b0e6/include/swift-c/SyntaxParser/SwiftSyntaxParser.h#L58

**Review by akyrtzi — CHANGES_REQUESTED** [review]
I'd recommend additional enhancement; if the parser library version is not specifically set with the new option and CMake variable, it should fallback to picking up the version from the compiler/toolchain version string, see here for the example from sourcekitd using that version string: https://github.com/apple/swift/blob/b5299531d4e498002920ebb0ddf853e12212b0e6/tools/SourceKit/CMakeLists.txt#L7

**Comment by nkcsgexi** [comment]
@akyrtzi Done with the toolchain version change.

**Review by akyrtzi — CHANGES_REQUESTED** [review]

**Review by akyrtzi — APPROVED** [review]

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by shahmishal** [comment]
@nkcsgexi We are seeing build failure in macOS toolchain due to this PR: 

```
00:48:49.008 ld: malformed 64-bit a.b.c.d.e version number: 4.2.2019020401
00:48:49.008 
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

https://ci.swift.org/job/oss-swift-package-osx/2922/console

**Comment by nkcsgexi** [comment]
This seems to be because the given version number is invalid. I'll take a look.

**Comment by nkcsgexi** [comment]
@shahmishal this other PR should fix the toolchain build failure: https://github.com/apple/swift/pull/22393

**Pull request merged**